### PR TITLE
CI: update to ROCm 6.0.2 and test MI300

### PR DESCRIPTION
**Pull request by fxmarty**
As per title

A few tests are failing due to `torch<2.2` and will be fixed once we bump to torch 2.3 + rocm6.0
```
FAILED tests/test_pipeline_mixin.py::FillMaskPipelineTests::test_fp16_casting - RuntimeError: "softmax_lastdim_kernel_impl" not implemented for 'Half'
FAILED tests/pipelines/test_pipelines_fill_mask.py::FillMaskPipelineTests::test_fp16_casting - RuntimeError: "softmax_lastdim_kernel_impl" not implemented for 'Half'
FAILED tests/models/mamba/test_modeling_mamba.py::MambaIntegrationTests::test_simple_generate_1_cpu - RuntimeError: "addmm_impl_cpu_" not implemented for 'Half'
FAILED tests/models/mamba/test_modeling_mamba.py::MambaIntegrationTests::test_simple_generate_cuda_kernels_tiny_1_cpu - RuntimeError: "addmm_impl_cpu_" not implemented for 'Half'
```

This PR requires a self hosted runner on MI300 first.

Report callbacks are skipped from the trainer tests, as codecarbon is not supported on ROCm and report callbacks are anyway tested independently (in `test_trainer_callback.py`).

### CODE REVIEW

File: tests/trainer/test_trainer_seq2seq.py [hunk 1]
```diff
@@ -119,6 +119,7 @@ def _compute_metrics(pred):
             warmup_steps=0,
             eval_steps=2,
             logging_steps=2,
+            report_to="tensorboard",
```
Comment (fxmarty) [inline tests/trainer/test_trainer_seq2seq.py:122]:
```suggestion
            report_to="none",
```

File: tests/trainer/test_trainer_seq2seq.py [hunk 2]
```diff
@@ -152,14 +153,15 @@ def test_return_sequences(self):
...
-        training_args = Seq2SeqTrainingArguments(".", predict_with_generate=True)
+        training_args = Seq2SeqTrainingArguments(".", predict_with_generate=True, report_to="tensorboard")
 
         trainer = Seq2SeqTrainer(
             model=model,
             args=training_args,
             tokenizer=tokenizer,
             data_collator=data_collator,
             compute_metrics=lambda x: {"samples": x[0].shape[0]},
+            report_to="tensorboard",
```
Comment (fxmarty) [inline tests/trainer/test_trainer_seq2seq.py:164]:
```suggestion
            report_to="none",
```

File: tests/trainer/test_trainer_seq2seq.py [hunk 3]
```diff
@@ -191,7 +193,9 @@ def test_bad_generation_config_fail_early(self):
         data_collator = DataCollatorForSeq2Seq(tokenizer, model=model, return_tensors="pt", padding="longest")
         gen_config = GenerationConfig(do_sample=False, top_p=0.9)  # bad: top_p is not compatible with do_sample=False
 
-        training_args = Seq2SeqTrainingArguments(".", predict_with_generate=True, generation_config=gen_config)
+        training_args = Seq2SeqTrainingArguments(
+            ".", predict_with_generate=True, generation_config=gen_config, report_to="tensorboard"
```
Comment (fxmarty) [inline tests/trainer/test_trainer_seq2seq.py:197]:
```suggestion
            ".", predict_with_generate=True, generation_config=gen_config, report_to="none"
```

File: docker/transformers-pytorch-amd-gpu/Dockerfile [hunk 4]
```diff
@@ -1,24 +1,20 @@
-FROM rocm/dev-ubuntu-20.04:5.6
+FROM rocm/pytorch:rocm6.0.2_ubuntu22.04_py3.10_pytorch_2.1.2
```
Comment (fxmarty) [inline docker/transformers-pytorch-amd-gpu/Dockerfile:1]: Let's update next week to PyTorch 2.3

**Comment by mht-sharma** [comment]
> As per title
> 
> A few tests are failing due to `torch<2.2` and will be fixed once we bump to torch 2.3 + rocm6.0
> 
> ```
> FAILED tests/test_pipeline_mixin.py::FillMaskPipelineTests::test_fp16_casting - RuntimeError: "softmax_lastdim_kernel_impl" not implemented for 'Half'
> FAILED tests/pipelines/test_pipelines_fill_mask.py::FillMaskPipelineTests::test_fp16_casting - RuntimeError: "softmax_lastdim_kernel_impl" not implemented for 'Half'
> FAILED tests/models/mamba/test_modeling_mamba.py::MambaIntegrationTests::test_simple_generate_1_cpu - RuntimeError: "addmm_impl_cpu_" not implemented for 'Half'
> FAILED tests/models/mamba/test_modeling_mamba.py::MambaIntegrationTests::test_simple_generate_cuda_kernels_tiny_1_cpu - RuntimeError: "addmm_impl_cpu_" not implemented for 'Half'
> ```
> 
> This PR requires a self hosted runner on MI300 first.
> 
> Report callbacks are skipped from the trainer tests, as codecarbon is not supported on ROCm and report callbacks are anyway tested independently (in `test_trainer_callback.py`).

With torch==2.3 RC, from above 4 the following tests passed:

```
tests/test_pipeline_mixin.py::FillMaskPipelineTests::test_fp16_casting 
tests/pipelines/test_pipelines_fill_mask.py::FillMaskPipelineTests::test_fp16_casting
tests/models/mamba/test_modeling_mamba.py::MambaIntegrationTests::test_simple_generate_cuda_kernels_tiny_1_cpu
```

The third test failed with:

```
FAILED tests/models/mamba/test_modeling_mamba.py::MambaIntegrationTests::test_simple_generate_1_cpu - AssertionError: Tensor-likes are not close!

```

**Comment by ydshieh** [comment]
I will review tomorrow.

@fxmarty Do we already have a MI300 runner?

**Review by ydshieh — APPROVED** [review]
Thank you @fxmarty for this PR. LGTM.

I will resolve the conflict later today or tomorrow.

For transparency, the (nvidia) daily CI  workflow file(s) has/have many changes in the past few month, and I haven't applied the same changes to AMD workflow files.

I will do that next week! But this PR is fine to be merged. I just want to know if the docker image could be built.

**Comment by ydshieh** [comment]
conflicts resolved

File: tests/extended/test_trainer_ext.py [hunk 5]
```diff
@@ -146,6 +146,7 @@ def test_trainer_log_level_replica(self, experiment_id):
             "do_eval": False,
             "do_predict": False,
             "n_gpus_to_use": 2,
+            "report_to": "none",
```
Comment (mht-sharma) [inline tests/extended/test_trainer_ext.py:149]: This argument is not required. The test fails with no such argument exist

**Comment by ydshieh** [comment]
Hi @mht-sharma 

Is anything I should wait before I merge this PR?

File: .github/workflows/self-scheduled-amd-mi300-caller.yml [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    branches: ["main"]
+    types: [completed]
+  push:
+    branches:
+      - run_amd_scheduled_ci_caller*
+      - mi300-ci
+
+jobs:
+  build-docker-containers:
+    # TODO: remove this
```
Comment (ydshieh) [inline .github/workflows/self-scheduled-amd-mi300-caller.yml:15]: @fxmarty Before I merge, could you remove this and any other places that should be removed if any.

Thanks!

Comment (ydshieh) [reply .github/workflows/self-scheduled-amd-mi300-caller.yml:15]: gently ping @fxmarty  🙏

**Comment by mht-sharma** [comment]
> Hi @mht-sharma
> 
> Is anything I should wait before I merge this PR?

Good from my side

**Comment by ydshieh** [comment]
I have removed the `TODO: remove this`.

I will merge as the PR is currently. They will need some updates however to match the recent changes on the nvidia daily CI workflow files.

**Comment by ydshieh** [comment]
Thank you for all the work you have done @fxmarty and @mht-sharma

**Pull request merged**